config/initializers/new_framework_defaults_7_1.rb

```diff
@@ -165,7 +165,7 @@
 # not yet been upgraded must be able to read caches from upgraded servers,
 # leave this optimization off on the first deploy, then enable it on a
 # subsequent deploy.
-# Rails.application.config.active_record.marshalling_format_version = 7.1
+Rails.application.config.active_record.marshalling_format_version = 7.1
```

Comment:
Does setting it this way actually work?

I noticed on a different issue that it seemed not to and worked around that - https://github.com/mastodon/mastodon/pull/27602/files#diff-692253a9e3680f63fd9a2ce0eb5e924bd2f4e9cd0983bf7bc6d89f1dc379080eR170

Based on https://github.com/rails/rails/issues/49827 my guess is that we have some gem interacting with this and we may need to set it directly on `ActiveRecord` for now (or other workaround?).

Reply:
Interesting. I applied your workaround.

Reply:
Good catch. Setting it directly in `application.rb` as follows works:
```patch
diff --git a/config/application.rb b/config/application.rb
index 990a89383e..02f8d56445 100644
--- a/config/application.rb
+++ b/config/application.rb
@@ -81,6 +81,8 @@ module Mastodon
     # config.paths.add File.join('app', 'api'), glob: File.join('**', '*.rb')
     # config.autoload_paths += Dir[Rails.root.join('app', 'api', '*')]
 
+    config.active_record.marshalling_format_version = 7.1
+
     config.active_job.queue_adapter = :sidekiq
 
     config.action_mailer.deliver_later_queue_name = 'mailers'
```

We do not use `delayed_job_active_record` but I suppose we could have a similar issue.

Reply:
Yes - I think when we eventually work through the full new framework settings update and have `7.1` as the defaults in application.rb the load order thing will become not relevant here, but it matters if we're setting it in that initializer.

And yeah - must be another gem ... I'm not sure how to trace down which one.

Reply:
I'm not sure either, I tried to track it down but failed.

Reply:
> Interesting. I applied your workaround.

Did you perhaps forget to push? I can't see that change.

Reply:
Kinda, my push failed...

Reply:
Pushed again now. I'll have a look at which dependency might be causing the issue.

Reply:
Thanks a lot!

Reply:
Alright, the problematic gem is `rails-settings-cached`. I can do a PR but it seems mastodon is using an old modified version. I'll see if I can untangle that.

Reply:
I reverted the workaround and pointed to the `rails-settings` PR instead.

Reply:
Unfortunately, this seems to break `rails-settings` functionality, and it doesn't seem to be enough to fix the issue we're facing.

Reply:
> Unfortunately, this seems to break `rails-settings` functionality,

Urk sorry, I'll have a look.

>  it doesn't seem to be enough to fix the issue we're facing.

How so? running `RAILS_ENV=development bin/rails runner 'p ActiveRecord::Marshalling.format_version'` give me `7.1` on this branch.

Reply:
Oh, that appears to work in the `production` and `development` environments, but not the `test` environment I was checking with!

Reply:
Ah, there might be another gem that breaks the test env, I can have a look.

Reply:
I think I'll just add the workaround back for now, so this can be merged.

Reply:
Ah I found the issue. It's because `RailsSettings::Settings` has `belongs_to :thing, polymorphic: true`. 

Since a couple Rails versions this assumes the relation is required and adds a validator by default, but since the gem was defining the model before Rails defaults are applied, it would use the old defaults...

I'll update the gem PR, but best course of action might be to just vendor the small parts of the gem that are needed and clean it up inside mastodon.

Reply:
> but not the test environment I was checking with!

Weird. `test` works for me.

Reply:
Not sure what is going on exactly:
```console
[mastodon@mastodev live]$ RAILS_ENV=development bin/rails runner 'p ActiveRecord::Marshalling.format_version' 
7.1
[mastodon@mastodev live]$ RAILS_ENV=test bin/rails runner 'p ActiveRecord::Marshalling.format_version' 
6.1
```

How are you tracking down those gems?

Reply:
> How are you tracking down those gems?

It's a bit finicky. I `bundle open activesupport`, then in `lazy_load_hooks.rb` I edit the `run_load_hooks` method:

```ruby
def run_load_hooks(name, base = Object)
  if name == :active_record
    raise "You naughty boy!"
  end
  # ...
end
```

This gives me a nice backtrace to the first code that trigger loading, until that is Rails itself.

Reply:
Ok, seems like the first thing to load `ActiveRecord` is `database_cleaner-active_record` then!